GeometryCore/src/LinearAlgebra.cpp

```diff
@@ -4,29 +4,107 @@
+            (intersection.y >= std::min(line2.a.y, line2.b.y)) && (intersection.y <= std::max(line2.a.y, line2.b.y)))
+        {
+            return true;
+        }
+        return false;
     }
 
     // Return distance from the p to the line
-    FPoint2D::coord distanceFromPointToLine(const FPoint2D& p, const LineSegment2D& line)
-    {
-        return {};
+    FPoint2D::coord distanceFromPointToLine(const FPoint2D& p, const LineSegment2D& line) {
+        auto projection = pointProjectionToLineSegment(p, line);
+
+        if ((projection.x >= std::min(line.a.x, line.b.x)) && (projection.x <= std::max(line.a.x, line.b.x)) &&
+            (projection.y >= std::min(line.a.y, line.b.y)) && (projection.y <= std::max(line.a.y, line.b.y))) {
+            return distance(projection, p);
+        }
+        else {
+            return std::min(distance(line.a, p), distance(line.b, p));
+        }
+    }
+
+    FPoint2D::coord distanceBetweenLineSegments(const LineSegment2D& line1, const LineSegment2D& line2) {
+        auto check_projection = [](const FPoint2D& p, const LineSegment2D& line, FPoint2D::coord& minimal_distance) { minimal_distance = std::min(distanceFromPointToLine(p, line), minimal_distance); };
+
+        FPoint2D::coord min_dis = std::min(std::min(distance(line1.a, line2.a), distance(line1.a, line2.b)), std::min(distance(line1.b, line2.a), distance(line1.b, line2.b)));
+        check_projection(line2.a, line1, min_dis);
+        check_projection(line2.b, line1, min_dis);
+        check_projection(line1.a, line2, min_dis);
+        check_projection(line1.b, line2, min_dis);
+
+        return min_dis;
     }
 
     // Return true if poly and line are colliding with each other.
     // Note: consider collision even if one of line end lies exactly on the polygon contour
-    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line)
-    {
-        return {};
+    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line) {
+        int size = poly.size();
+        for (int i = 0; i < size; ++i) {
+            LineSegment2D side(poly[i], poly[(i + 1) % size]);
+            if (lineSegmentsCollide(line, side)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    //Return true if polygon is close to line (closer than given gap)
+    bool polygonAndLineSegmentAreAdjacent(const Polygon& poly, const LineSegment2D& line, const FPoint2D::coord gap) {
+        int size = poly.size();
+        std::vector<FPoint2D::coord> distance_from_line_to_polygon_side(size);
+
+        #pragma omp parallel for
+        for (int i = 0; i < size; ++i) {
+                LineSegment2D side(poly[i], poly[(i + 1) % size]);
+                distance_from_line_to_polygon_side[i] = distanceBetweenLineSegments(line, side);
+        }
+        return *std::min_element(distance_from_line_to_polygon_side.begin(), distance_from_line_to_polygon_side.end()) <= gap;
```

Comment:
If poly is empty polygon, than here we will have undefined behavior (dereferencing of iterator not pointing to containers elements)